db/migrations/102_rework_enter_permissions.sql

```diff
@@ -0,0 +1,56 @@
+-- +migrate Up
+ALTER TABLE `permissions_granted`
+    ADD COLUMN `can_enter_from` DATETIME NOT NULL DEFAULT '9999-12-31 23:59:59'
+        COMMENT 'Time from which the group can “enter” this item' AFTER `can_view`,
+    ADD COLUMN `can_enter_until` DATETIME NOT NULL DEFAULT '9999-12-31 23:59:59'
+        COMMENT 'Time until which the group can “enter” this item' AFTER `can_enter_from`,
+    MODIFY COLUMN `can_grant_view` ENUM('none','enter','content','content_with_descendants','solution','solution_with_grant') NOT NULL DEFAULT 'none'
+        COMMENT 'The level of visibility that the group can give on this item to other groups on which it has the right to';
+
+ALTER TABLE `permissions_generated`
+    MODIFY COLUMN `can_grant_view_generated`
+        ENUM('none','enter','content','content_with_descendants','solution','solution_with_grant') NOT NULL DEFAULT 'none'
+        COMMENT 'The aggregated level of visibility that the group can give on this item to other groups on which it has the right to';
+
+ALTER TABLE `groups_contest_items`
+    DROP COLUMN can_enter_from,
+    DROP COLUMN can_enter_until;
+
+ALTER TABLE `items`
+    ADD COLUMN `entering_time_min` DATETIME DEFAULT NULL
+        COMMENT 'Lower bound on the entering time. Has the priority over given can_enter_from/until permissions.'
+        AFTER `contest_entering_condition`,
+    ADD COLUMN `entering_time_max` DATETIME DEFAULT NULL
+        COMMENT 'Upper bound on the entering time. Has the priority over given can_enter_from/until permissions.'
+        AFTER `entering_time_min`;
```

Comment:
Contrary to what I said, let's consider these two attributes NOT NULL. It will be more consistent with `contest_entering_condition`, `contest_max_team_size`, ... and just use the default values when not applicable. Handling null values seems more complicated and less clear.

Reply:
fixed